rust/connlib/tunnel/src/client.rs

```diff
@@ -950,6 +858,100 @@ impl ClientState {
     pub(crate) fn poll_transmit(&mut self) -> Option<snownet::Transmit<'_>> {
         self.node.poll_transmit()
     }
+
+    pub fn set_resources(&mut self, resources: &[ResourceDescription]) {
+        self.remove_resources(
+            &HashSet::<ResourceId>::from_iter(resources.iter().map(|r| r.id()))
+                .difference(&HashSet::from_iter(self.resource_ids.keys().copied()))
+                .copied()
+                .collect_vec(),
+        );
+
+        self.add_resources(
+            &HashSet::<ResourceDescription>::from_iter(self.resource_ids.values().cloned())
+                .difference(&HashSet::from_iter(resources.iter().cloned()))
+                .cloned()
+                .collect_vec(),
+        );
+    }
+
+    pub fn add_resources(&mut self, resources: &[ResourceDescription]) {
+        for resource_description in resources {
+            if let Some(resource) = self.resource_ids.get(&resource_description.id()) {
+                if resource.has_different_address(resource) {
+                    self.remove_resource(resource.id());
+                }
+            }
+
+            match &resource_description {
+                ResourceDescription::Dns(dns) => {
+                    self.dns_resources.insert(dns.address.clone(), dns.clone());
+                }
+                ResourceDescription::Cidr(cidr) => {
+                    self.cidr_resources.insert(cidr.address, cidr.clone());
+                }
+            }
+
+            self.resource_ids
+                .insert(resource_description.id(), resource_description.clone());
+        }
+    }
+
+    fn remove_resources(&mut self, ids: &[ResourceId]) {
+        for id in ids {
+            self.remove_resource(*id);
+        }
+    }
```

Comment:
Nit: Is this function worth having? I think we can just iterate at the call-site here.